dig_test.go

```diff
@@ -627,6 +629,69 @@ func TestEndToEndSuccess(t *testing.T) {
 		}), "both objects should be successfully resolved on Invoke")
 	})
 
+	t.Run("struct constructor with as interface option", func(t *testing.T) {
+		c := New()
+
+		provider := c.Provide(
+			func() *bytes.Buffer {
+				var buf bytes.Buffer
+				buf.WriteString("foo")
+				return &buf
+			},
+			As(new(fmt.Stringer), new(io.Reader)),
+		)
+
+		require.NoError(t, provider, "provide failed")
+
+		require.NoError(t, c.Invoke(
+			func(s fmt.Stringer, r io.Reader) {
+				require.Equal(t, "foo", s.String(), "invoke got new buffer")
+				got, err := ioutil.ReadAll(r)
+				assert.NoError(t, err, "failed to read from reader")
+				require.Equal(t, "foo", string(got), "invoke got new buffer")
+			},
+		), "invoke failed")
+	})
```

Comment:
I would add another check to this test to make sure there's no `*bytes.Buffer` in the container.

```
require.Error(t, c.Invoke(func(*bytes.Buffer) {
    t.Fatalf("must not be called")
}), "must not have a *bytes.Buffer in the container")
```